src/main/kotlin/org/rust/lang/core/resolve/indexes/RustImplIndex.kt

```diff
@@ -0,0 +1,221 @@
+            .filter { it.isStatic }
+
+    fun findMethodsFor(target: RustType, project: Project): Sequence<RustFnElement> =
+        findImplsFor(target, project)
+            .flatMap { it.implBody?.implMethodMemberList.orEmpty().asSequence() } +
+
+        findTraitsImplementedFor(target, project)
+            .flatMap { it.traitBody.traitMethodMemberList.orEmpty().asSequence() }
+
+    fun findTraitsImplementedFor(target: RustType, project: Project): Sequence<RustTraitItemElement> {
+        var traitRefs = emptySequence<RustTraitRefElement>()
+
+        if (target is RustTypeParameterType)
+            traitRefs += target.parameter.bounds.mapNotNull { it.bound.traitRef }
+        else
+            traitRefs += findImplsFor(target, project).mapNotNull { it.traitRef}
+
+        return traitRefs.mapNotNull { it.path.reference.resolve() as? RustTraitItemElement }
+    }
+
+
+    fun findImplsFor(target: RustType, project: Project): Sequence<RustImplItemElement> {
+        var inherentImpls = emptySequence<RustImplItemElement>()
+        if (target is RustStructOrEnumTypeBase)
+            inherentImpls = findInherentImplsForInternal(target.item)
+
+        return findNonInherentImplsForInternal(target.decay, project)
+                    .filter {
+                        impl -> impl.type?.let { it.resolvedType == target } ?: false
+                    } + inherentImpls
+    }
+
+    private fun findInherentImplsForInternal(target: RustStructOrEnumItemElement): Sequence<RustImplItemElement> {
+        val found = arrayListOf<RustImplItemElement>()
+
+        val aliases = arrayListOf(target.name!!)
+
+        StubIndex
+            .getInstance()
+            .processElements(
+                RustAliasIndex.KEY,
+                target.name!!,
+                target.project,
+                GlobalSearchScope.allScope(target.project),
+                RustUseItemElement::class.java,
+                {
+                    it.name?.let {
+                        aliases.add(it)
+                    }
+
+                    true /* continue */
+                })
+
+
+        aliases.forEach { alias ->
+            StubIndex
+                .getInstance()
+                .processElements(
+                    ByName.KEY,
+                    alias,
+                    target.project,
+                    GlobalSearchScope.allScope(target.project),
+                    RustImplItemElement::class.java,
+                    {
+                        found.add(it)
+                        true /* continue */
+                    })
+        }
+
+        return found.asSequence()
+                    .filter { impl ->
+                        impl.type?.let {
+                            it.resolvedType.let { ty ->
+                                ty is RustStructOrEnumTypeBase &&
+
+                                // TODO(XXX): Revisit
+                                ty.item.canonicalCratePath == target.canonicalCratePath
```

Comment:
Didn't give it much of attention due to the fact that, though after parse being triggered stub-based PSI nodes may be still there due to outstanding references being held still.